File: src/wall_follow_node/src/wall_follow_node.cpp
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/laser_scan.hpp"
#include "ackermann_msgs/msg/ackermann_drive_stamped.hpp"
#include <cmath>
#include <limits>

class WallFollowNode : public rclcpp::Node {
public:
  WallFollowNode()
  : Node("wall_follow_node"),
    // Parameters as in the F1Tenth Lab3 template
    desired_distance_(1.0), // Desired distance from the right wall [meters]
    k_(0.8),               // Proportional gain for steering control
    L_(1.0),               // Lookahead distance [meters]
    speed_(20.0),           // Constant forward speed [m/s]
    phi_(M_PI/4)           // Angle offset (45° in radians) for the first measurement
  {
    // Subscribe to the laser scan topic (usually "/scan")
    scan_sub_ = this->create_subscription<sensor_msgs::msg::LaserScan>(
      "/scan", 10, std::bind(&WallFollowNode::scan_callback, this, std::placeholders::_1)
    );

    // Publisher for drive commands (usually "/drive")
    drive_pub_ = this->create_publisher<ackermann_msgs::msg::AckermannDriveStamped>(
      "/drive", 10
    );

    RCLCPP_INFO(this->get_logger(), "Wall Follow Node Initialized");
  }

private:
  // Parameters
  double desired_distance_;
  double k_;
  double L_;
  double speed_;
  double phi_;  // Offset angle (radians) for the first laser measurement

  // ROS interfaces
  rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr scan_sub_;
  rclcpp::Publisher<ackermann_msgs::msg::AckermannDriveStamped>::SharedPtr drive_pub_;

  // Callback processing the laser scan message for wall following
  void scan_callback(const sensor_msgs::msg::LaserScan::ConstSharedPtr scan_msg) {
    // Define the angles (in radians) for the two measurements:
    // b: directly to the right (–90°)
    // a: offset by φ from the right (i.e. –90° + φ)
    double angle_b = -M_PI / 2.0;
    double angle_a = angle_b + phi_;

    // Verify that these angles are within the laser scan range.
    if (angle_a < scan_msg->angle_min || angle_b < scan_msg->angle_min ||
        angle_a > scan_msg->angle_max || angle_b > scan_msg->angle_max) {
      RCLCPP_WARN(this->get_logger(), "Desired laser angles (%.2f, %.2f) out of scan range (%.2f, %.2f).",
                  angle_a, angle_b, scan_msg->angle_min, scan_msg->angle_max);
      return;
    }

    // Calculate the corresponding indices in the scan array.
    int index_b = static_cast<int>((angle_b - scan_msg->angle_min) / scan_msg->angle_increment);
    int index_a = static_cast<int>((angle_a - scan_msg->angle_min) / scan_msg->angle_increment);

    // Check index validity.
    if (index_a < 0 || index_a >= static_cast<int>(scan_msg->ranges.size()) ||
        index_b < 0 || index_b >= static_cast<int>(scan_msg->ranges.size())) {
      RCLCPP_WARN(this->get_logger(), "Calculated laser scan indices out of range.");
      return;
    }

    double b = scan_msg->ranges[index_b];
    double a = scan_msg->ranges[index_a];

    // Check that both measurements are valid.
    if (!std::isfinite(a) || !std::isfinite(b)) {
      RCLCPP_WARN(this->get_logger(), "Invalid laser scan measurements: a = %.2f, b = %.2f", a, b);
      return;
    }

    // Compute the wall’s angle (alpha) relative to the vehicle.
    // Formula: alpha = arctan((a*cos(phi) - b) / (a*sin(phi)))
    double alpha = std::atan((a * std::cos(phi_) - b) / (a * std::sin(phi_)));

    // Current distance from the wall (projection using measurement b)
    double D_t = b * std::cos(alpha);
    // Predicted (projected) distance from the wall after moving ahead by lookahead distance L_
    double D_t1 = D_t + L_ * std::sin(alpha);

    // Compute the error between the desired distance and the projected distance.
    double error = desired_distance_ - D_t1;

    // Compute the steering angle command using a proportional controller.
    double steering_angle = k_ * error;

    // Build and publish the drive command.
    ackermann_msgs::msg::AckermannDriveStamped drive_msg;
    drive_msg.header.stamp = this->now();
    drive_msg.drive.speed = speed_;
    drive_msg.drive.steering_angle = steering_angle;
    drive_pub_->publish(drive_msg);

    // Log the values for debugging.
    RCLCPP_INFO(this->get_logger(), 
                "Laser measurements: a = %.2f, b = %.2f, alpha = %.2f, D_t = %.2f, D_t1 = %.2f, error = %.2f, steer = %.2f",
                a, b, alpha, D_t, D_t1, error, steering_angle);
  }
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<WallFollowNode>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

File: src/wall_follow_node/scripts/wall_follow_node.py
#!/usr/bin/env python3
import math
import numpy as np
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import LaserScan
from ackermann_msgs.msg import AckermannDriveStamped

# PID control parameters
kp = 1.5
kd = 0.05
ki = 0.25
servo_offset = 0.0
prev_error = 0.0 
integral = 0.0
prev_time = 0.0

# Wall follow parameters
ANGLE_RANGE = 270              # Hokuyo 10LX has 270° scan
DESIRED_DISTANCE_RIGHT = 0.9   # meters
DESIRED_DISTANCE_LEFT = 0.85   # meters
VELOCITY = 5.0                 # m/s
CAR_LENGTH = 1.0               # meters

class WallFollow(Node):
    def __init__(self):
        super().__init__('wall_follow_node')
        global prev_time
        # Initialize time using the node's clock
        prev_time = self.get_clock().now().nanoseconds / 1e9

        # Subscribe to the LIDAR scan topic
        self.lidar_sub = self.create_subscription(
            LaserScan,
            '/scan',
            self.lidar_callback,
            10
        )

        # Publisher for drive commands (publish on /nav as per your configuration)
        self.drive_pub = self.create_publisher(
            AckermannDriveStamped,
            '/drive',
            10
        )

        self.get_logger().info("WallFollow node initialized.")

    def get_range(self, data, angle):
        """
        Get distance reading for a given angle from LIDAR data.
        Angle is in degrees (from -45 to 225, where 0° is directly to the right).
        """
        if angle < -45 or angle > 225:
            return float('nan')
        # Map the angle to an index in the data array.
        # Assuming the LIDAR provides a 360° reading (or a subset mapped accordingly).
        index = int(len(data) * (angle + 90) / 360)
        if index < 0 or index >= len(data):
            return float('nan')
        if not np.isnan(data[index]) and not np.isinf(data[index]):
            return data[index]
        return float('nan')

    def pid_control(self, error, velocity):
        global integral, prev_error, kp, ki, kd, prev_time
        current_time = self.get_clock().now().nanoseconds / 1e9  # seconds
        dt = current_time - prev_time if (current_time - prev_time) > 0 else 1e-3
        integral += error * dt
        derivative = (error - prev_error) / dt
        angle = kp * error + ki * integral + kd * derivative
        prev_error = error
        prev_time = current_time

        drive_msg = AckermannDriveStamped()
        drive_msg.header.stamp = self.get_clock().now().to_msg()
        drive_msg.header.frame_id = "laser"
        drive_msg.drive.steering_angle = -angle + servo_offset

        abs_angle_deg = abs(math.degrees(angle))
        if abs_angle_deg <= 10:
            drive_msg.drive.speed = velocity
        elif abs_angle_deg <= 20:
            drive_msg.drive.speed = 1.0
        else:
            drive_msg.drive.speed = 0.5

        self.drive_pub.publish(drive_msg)

    def follow_left(self, data, left_dist):
        """
        Calculate error between desired and actual distance from left wall.
        Uses two LIDAR measurements.
        """
        front_scan_angle = 125  # degrees
        back_scan_angle = 180   # degrees
        teta = math.radians(abs(front_scan_angle - back_scan_angle))
        front_scan_dist = self.get_range(data, front_scan_angle)
        back_scan_dist = self.get_range(data, back_scan_angle)
        if np.isnan(front_scan_dist) or np.isnan(back_scan_dist):
            return 0.0
        alpha = math.atan2(front_scan_dist * math.cos(teta) - back_scan_dist,
                             front_scan_dist * math.sin(teta))
        wall_dist = back_scan_dist * math.cos(alpha)
        ahead_wall_dist = wall_dist + CAR_LENGTH * math.sin(alpha)
        return left_dist - ahead_wall_dist

    def lidar_callback(self, msg):
        error_val = self.follow_left(msg.ranges, DESIRED_DISTANCE_LEFT)
        self.pid_control(error_val, VELOCITY)

def main(args=None):
    rclpy.init(args=args)
    node = WallFollow()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

